src/tools/schema2graph.ts

```diff
@@ -0,0 +1,144 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {Schema} from '../runtime/schema.js';
+import {Direction} from '../runtime/manifest-ast-nodes.js';
+import {Dictionary} from '../runtime/hot.js';
+import {Manifest} from '../runtime/manifest.js';
+
+export interface FieldEntry {
+  particleName: string;
+  schema: Schema;
+  direction: Direction;
+  connectionName: string;
+}
+
+export interface TypeLattice {
+  nodes: Dictionary<FieldEntry>;
+  edges: Dictionary<string[]>;
+  nodesByParticle: Dictionary<FieldEntry[]>;
+  contains(name: string): boolean;
+}
+
+class Builder {
+  manifestData: Iterable<FieldEntry> = null;
+
+  public from(manifest: Manifest): Builder {
+    this.manifestData = this.parse(manifest);
+    return this;
+  }
+
+  public to(graph: TypeGraph) {
+    if (this.manifestData === null) {
+      throw new Error('Call `from` on a manifest first!');
+    }
+    this.buildGraph(this.manifestData, graph);
+  }
+
+  public build(): TypeLattice {
+    const graph = new TypeGraph();
+    this.to(graph);
+    return graph;
+  }
+
+  protected* parse(manifest: Manifest): Iterable<FieldEntry> {
+    for (const particle of manifest.allParticles) {
+      for (const connection of particle.connections) {
+        const schema: Schema = connection.type.getEntitySchema();
+        if (!schema) {
+          continue;
+        }
+        yield {
+          particleName: particle.name,
+          schema,
+          direction: connection.direction,
+          connectionName: connection.name,
+        };
+      }
+    }
+  }
+
+  protected buildGraph(entries: Iterable<FieldEntry>, graph: TypeGraph) {
+    const allowedDirections = ['in', 'inout', 'out'];
+    const isHandleDirection = (e: FieldEntry): boolean => allowedDirections.includes(e.direction);
+
+    for (const entry of entries) {
+      if (isHandleDirection(entry)) {
+        graph.addNode(entry);
+      }
+    }
+    for (const particleEntries of Object.values(graph.nodesByParticle)) {
+      for (const e0 of particleEntries) {
+        for (const e1 of particleEntries) {
+          if (!isHandleDirection(e0) || !isHandleDirection(e1)) {
+            continue;
+          }
+          const source = nameFromEntry(e0);
+          const name = nameFromEntry(e1);
+          if (source !== name && e0.schema.isMoreSpecificThan(e1.schema)) {
+            graph.addEdge(source, name);
+          }
+        }
+      }
+    }
+  }
+}
+
+export function nameFromEntry(entry: FieldEntry): string {
+  return `${entry.particleName}_${entry.connectionName}`;
+}
+
+export class TypeGraph implements TypeLattice {
+  edges: Dictionary<string[]>;
+  nodes: Dictionary<FieldEntry>;
```

Comment:
This feels like an odd way to structure this graph. I would expect a node class that has a list of its parent (and/or children?) nodes, with a separate set holding the roots. That should be more amenable to later traversal logic.

Reply:
This, and the Builder look roughly similar to the way Guava's Graph/GraphBuilder API looks, in the sense that it doesn't impose a contract on the nodes of the graph, but anything with equals()/hashCode() works. I don't think there's anything wrong with doing it this way, but I'd argue if you're going to have a 'general purpose' graph by storing edges and vertices externally, you should probably actually make it general, or look for an existing typescript library that provides such a graph.